File: ios/Classes/Api/PaymentWithCardTokenApi.swift
//
//  PaymentWithCardTokenApi.swift
//  pgw_sdk
//
//  Created by Nonthawatt Phongwittayapanu on 9/2/2565 BE.
//

import Foundation
import PGW

class PaymentWithCardToken {
    private var call: FlutterMethodCall
    private var result: FlutterResult
    
    init(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        self.call = call
        self.result = result
    }
    
    func proceed() {
        guard let dicts = call.arguments as? [String: Any],
              let jsonString = dicts["request"] as? String
        else {
            result(FlutterError.init(code: "998", message: "Arguments exception", details: nil))
            return
        }
        guard let request = mappingRequest(json: jsonString) else {
            result(FlutterError.init(code: "900", message: "Json exception", details: nil))
            return
        }
        
        PGWSDK.shared.proceedTransaction(transactionResultRequest: request, { (response: TransactionResultResponse) in
            if response.responseCode == APIResponseCode.TransactionAuthenticateRedirect ||
                response.responseCode == APIResponseCode.TransactionAuthenticateFullRedirect ||
                response.responseCode == APIResponseCode.TransactionCompleted {
                
                let responseDict = [
                    "channelCode": response.channelCode as String,
                    "invoiceNo": response.invoiceNo as String,
                    "type": response.type as String,
                    "data": response.data as String,
                    "fallbackData": response.fallbackData as String,
                    "expiryTimer": response.expiryTimer as Int,
                    "expiryDescription": response.expiryDescription as String,
                    "responseCode": response.responseCode as String,
                    "responseDescription": response.responseDescription as String,
                ] as [String : Any]
                
                
                do {
                    let data = try JSONSerialization.data(withJSONObject: responseDict)
                    if let string = String(data: data, encoding: .utf8) {
                        print(string)
                        return self.result(string)
                    }
                } catch let error as NSError {
                    print(error)
                }
                //            } else if response.responseCode == APIResponseCode.TransactionCompleted {
                //                //Inquiry payment result by using invoice no.
            } else {
                //Get error response and display error.
                self.result(FlutterError.init(code: response.responseCode, message: response.responseDescription, details: nil))
            }
        }) { (error: NSError) in
            //Get error response and display error.
            self.result(FlutterError.init(code: "\(error.code)", message: error.localizedDescription, details: nil))
        }
    }
    
    private func mappingRequest(json: String) -> TransactionResultRequest? {
        do {
            let jsonData = json.data(using: .utf8)!
            let jsonDict = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any?]
            let paymentToken = jsonDict?["paymentToken"] as? String ?? ""
            
            /// Get "payment": {...}
            let jsonPayment = jsonDict?["payment"] as? [String: Any?] ?? [:]
            
            /// Get "code": {...}, "data": {...}
            let jsonPaymentCode = jsonPayment["code"] as? [String: Any?] ?? [:]
            let jsonPaymentData = jsonPayment["data"] as? [String: Any?] ?? [:]
            
            let paymentCode: PaymentCode = PaymentCode(
                channelCode: jsonPaymentCode["channelCode"] as? String ?? "",
                jsonPaymentCode["agentCode"] as? String ?? "",
                jsonPaymentCode["agentChannelCode"] as? String ?? ""
            )
            guard let cardToken = jsonPaymentData["token"] as? String else {
                return nil
            }
            guard let securityCode = jsonPaymentData["securityCode"] as? String else {
                return nil
            }
            
            let paymentRequest = CardTokenPaymentBuilder(paymentCode: paymentCode, cardToken).securityCode(securityCode).build()
            var builder = TransactionResultRequestBuilder(paymentToken: paymentToken)
            if let clientId = jsonDict?["clientID"] as? String {
                builder = builder.clientId(clientId)
            }
            if let locale = jsonDict?["locale"] as? String {
                builder = builder.locale(locale)
            }
            
            return builder
                .with(paymentRequest)
                .build()
        } catch let error as NSError {
            print(error)
            return nil
        }
    }
    
}

File: android/src/main/kotlin/com/ccpp/pgw_sdk/api/PaymentWithCardTokenApi.kt
/*
 * Created by Nonthawatt Phongwittayapanu on 2/8/22, 1:49 PM
 * Copyright © 2022 2C2P. All rights reserved.
 */

package com.ccpp.pgw_sdk.api

import android.util.Log
import androidx.annotation.NonNull
import com.ccpp.pgw.sdk.android.builder.CardTokenPaymentBuilder
import com.ccpp.pgw.sdk.android.builder.TransactionResultRequestBuilder
import com.ccpp.pgw.sdk.android.callback.APIResponseCallback
import com.ccpp.pgw.sdk.android.core.PGWSDK
import com.ccpp.pgw.sdk.android.enums.APIResponseCode
import com.ccpp.pgw.sdk.android.model.PaymentCode
import com.ccpp.pgw.sdk.android.model.api.TransactionResultRequest
import com.ccpp.pgw.sdk.android.model.api.TransactionResultResponse
import com.ccpp.pgw_sdk.helper.JsonHelper
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

class PaymentWithCardTokenApi(
    @NonNull val call: MethodCall, @NonNull val result: MethodChannel.Result
) {

    /**
     * Proceed function
     */
    fun request() {
        val json = call.argument<String>("request") ?: ""
        val request = transactionRequestFromJson(json) ?: kotlin.run {
            result.error("998", "Parse json error", null)
            return
        }

        PGWSDK.getInstance()
            .proceedTransaction(request, object : APIResponseCallback<TransactionResultResponse> {
                override fun onResponse(response: TransactionResultResponse) {
                    when (response.responseCode) {
                        APIResponseCode.TransactionQRPayment,
                        APIResponseCode.TransactionPaymentSlip,
                        APIResponseCode.TransactionAuthenticateRedirect,
                        APIResponseCode.TransactionAuthenticateFullRedirect -> {
                            val redirectUrl = response.data //Open WebView
                            Log.d("DEMO", "onResponse: =>> $response")
                            Log.d("DEMO", "onResponse: Url =>> $redirectUrl")
                            result.success(transactionResponseToJson(response))
                        }
                        APIResponseCode.TransactionCompleted -> {
                            Log.d("DEMO", "onResponse: Transaction Complete")
                            result.success(transactionResponseToJson(response))
                        }
                        else -> {
                            //Get error response.
                            Log.d("DEMO", "onResponse: Error ${response.responseCode}")
                            Log.d("DEMO", "onResponse: Error ${response.responseDescription}")

                            result.error(response.responseCode, response.responseDescription, null)
                        }
                    }
                }

                override fun onFailure(error: Throwable) {

                    //Get error response and display error.
                    result.error("999", error.localizedMessage, error)
                }
            })
    }

    private fun transactionRequestFromJson(json: String): TransactionResultRequest? {
        val jsonObject = JSONObject(json)

        Log.d("DEMO", "parseJson: Request =>> ${jsonObject.toString(2)}")

        val paymentToken = jsonObject["paymentToken"].toString()
        val clientId = JsonHelper.jsonString(jsonObject, "clientID")
        val locale = JsonHelper.jsonString(jsonObject, "locale")
        return try {
            val jsonPayment = jsonObject["payment"] as JSONObject// JSON payment
            val jsonPaymentCode = jsonPayment["code"] as JSONObject
            val jsonPaymentData = jsonPayment["data"] as JSONObject

            val cardToken = JsonHelper.jsonString(jsonPaymentData, "token")
            val securityCode = JsonHelper.jsonString(jsonPaymentData, "securityCode")

            val code = PaymentCode(
                JsonHelper.jsonString(jsonPaymentCode, "channelCode"),
                JsonHelper.jsonString(jsonPaymentCode, "agentCode"),
                JsonHelper.jsonString(jsonPaymentCode, "agentChannelCode"),
            )

            val paymentRequest = CardTokenPaymentBuilder(code, cardToken).setSecurityCode(securityCode).build()

            return TransactionResultRequestBuilder(paymentToken).apply {
                if (!clientId.isNullOrBlank()) {
                    setClientId(clientId)
                }
                if (!locale.isNullOrBlank()) {
                    setLocale(locale)
                }
                with(paymentRequest)
            }.build()
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    private fun transactionResponseToJson(response: TransactionResultResponse): String {
        val responseMap = mapOf<String, Any>(
            "channelCode" to response.channelCode,
            "invoiceNo" to response.invoiceNo,
            "type" to response.type,
            "data" to response.data,
            "fallbackData" to response.fallbackData,
            "expiryTimer" to response.expiryTimer,
            "expiryDescription" to response.expiryDescription,
            "responseCode" to response.responseCode,
            "responseDescription" to response.responseDescription,
        )

        return JSONObject(responseMap).toString()
    }
}